src/PetProject.Framework.Kafka/Consumer/Consumer.cs

```diff
@@ -181,42 +172,68 @@ protected virtual void HandleOnConsumerError(object sender, Message message)
             {
-                var key = new StringDeserializer(Encoding.UTF8).Deserialize(message.Topic, message.Key);
-                var messageContent = new JsonDeserializer<MessageWrapper>().Deserialize(message.Topic, message.Value);
-
-                var report = producer.ProduceAsync(message.Topic, key, messageContent).Result;
+                var report = producer.ProduceAsync(topicName, key, messageValue).Result;
 
                 if (!report.Error.HasError)
                 {
+                    // move offset forward to this message's offset + 1
+                    this.confluentConsumer.CommitAsync(new[]
+                    {
+                        new TopicPartitionOffset(topicPartition, offset + 1L)
+                    }).Wait();
                     return;
                 }
 
-                this.logger.KafkaLogCritical("Critical Error when retrying to send message. Topic: {topic} | Message: {message} | Timestamp: {timestamp} | Error: {error}", message.Topic, messageContent, report.Timestamp, report.Error);
+                this.logger.KafkaLogCritical("Critical Error when retrying to send message. Topic: {topic} | Message: {message} | Timestamp: {timestamp} | Error: {error}", topicName, messageValue, report.Timestamp, report.Error);
             }
         }
 
+        protected void RequeueMessageOnError(Message<string, MessageWrapper> message)
+        {
+            this.RequeueMessageOnError(message.Topic, message.TopicPartition, message.Offset, message.Key, message.Value);
+        }
+
         /// <summary>
         /// May be overriden by the client in order to consumer messages and add custom treatment.
         /// </summary>
         /// <param name="sender">Message sender.</param>
         /// <param name="consumerMessage">Message envelope with key value pair content.</param>
         protected void HandleMessage(object sender, Message<string, MessageWrapper> consumerMessage)
         {
-            if (consumerMessage.Value == null)
+            try
             {
-                this.logger.KafkaLogInfo("ConsumerMessage has no content: {consumerMessage}", consumerMessage);
-                return;
-            }
-
-            var wrappedMessage = consumerMessage.Value;
+                if (consumerMessage.Value == null)
+                {
+                    this.logger.KafkaLogInfo("ConsumerMessage has no content: {consumerMessage}", consumerMessage);
+                }
+                else
+                {
+                    var wrappedMessage = consumerMessage.Value;
+                    this.CallHandler(wrappedMessage);
+                }
 
-            this.CallHandler(wrappedMessage);
+                this.confluentConsumer.CommitAsync(consumerMessage).Wait();
```

Comment:
You loose too much control. In this case you consume and commit immediately. What happens if the processing of the message fails. The offset already is committed and advanced in the partition. The next poll won't catch the message.

Reply:
If processing fails, i.e., handler throws an exception, then this comitasync won't be called.

Reply:
What you're saying is that we should expose the internal Confluent Message object to the user which for me doesn't make sense because why would we develop a consumer of our own if we don't abstract this kind of low level logic?

Reply:
If the user wants to have more control over these things he can just override the methods or maybe we can think of implementing a new consumer that is lower level than this one.

Reply:
We can expose a custom message object that maps to the confluent one. This object may be reused if we change the consumer tech below.

Reply:
Hmm that's a good idea but as a feature in the near future, not now 😄